aten/src/ATen/functorch/BatchRulesScatterOps.cpp

```diff
@@ -1123,20 +1124,48 @@ std::tuple<Tensor,optional<int64_t>> index_fill_int_tensor_batch_rule_impl(
   auto batch_size = get_bdim_size3(self, self_bdim, index, index_bdim, value, value_bdim);
   self_ = ensure_has_bdim(self_, self_bdim.has_value(), batch_size);
   index_ = ensure_has_bdim(index_, index_bdim.has_value(), batch_size);
-  value_ = ensure_has_bdim(value_, value_bdim.has_value(), batch_size);
+
+  if (inplace || value_bdim.has_value()) {
+    // Do for-loop for in-place because we cannot reshape
+    // `self_` having an incompatible stride without copying.
+    // If value has a batch dim, we do for-loop as well because
+    // index_fill_ supports 1-element tensor only.
+    for (const auto i : c10::irange(0, batch_size)) {
+      const auto& self_slice = self_.select(0, i);
+      const auto& index_slice = index_.select(0, i);
+      self_slice.index_fill_(
+        dim,
+        index_slice,
+        value_bdim.has_value() ? value_.select(0, i) : value_
```

Comment:
Can we write this for-loop such that we won't have to check `value_bdim` every iteration?

Reply:
I feel this is the most concise version, and checking `value_bdim` would not be costly.

Here are the ways I came up:

1. Always ensure `value` has a batch dim and do select in every iteration - costly.
2. Make copies of the for-loop statement and separate the work out. There are four combinations for `inplace` and `value_bdim.has_value()` - not concise.
3. Determine `value_bdim.has_value()` before for-loop and use lambdas to wrap either `value_` or `value_.select(0, i)` - I feel unnecessary tbh. Would it be better?

Or maybe I'm missing something.

Reply:
Hmm, makes sense, plus I would assume that most of the time would be spent in `index_fill` so this should be good.